### PR TITLE
Video-specific permissions

### LINKED ISSUE #341: Video-specific view permissions (backend)
Collection admins should be able to override the default collection view permissions for specific videos. 

Available video view permissions should include:
  - Private (collection owner and admins only)
  - Public (everyone who's allowed to log in to OVS)
  - Moira lists (these will override collection view lists but not collection admin lists)
  - None (default - the collection view permissions will be used to determine who can view)

### LINKED ISSUE #342: Display permissions options on video edit form
Depends on #341 

The edit form dialog for videos should include options for choosing who should be able to view the video.  Mirroring the backend, the options should include:

- 'Collection viewers': the default, meaning the collection view permissions will be used.
- 'Only me': only the collection admin(s)
- 'Everyone': everyone who's allowed to log in to OVS
- 'Moira lists': these will override the collection view lists but not the collection admin lists

Here is a preliminary screenshot (@robhouse what do you think?):

screen shot 2017-11-01 at 10 58 21 am

### PR DESCRIPTION
**Pull request by mbertrand**
#### What are the relevant tickets?
Closes #341 
Closes #342 

#### What's this PR do?
Adds video-specific permissions (including 'public') that override collection permissions.

#### How should this be manually tested?
- Set `ENABLE_VIDEO_PERMISSIONS=True` in your `.env` file.
  - Set collection view and admin permissions to certain levels (only me or moira lists) and verify that they still work as designed. (Superusers, collection owners and people on admin lists or view lists should be able to view the video).
  - Set video permissions to public.  Try to view in another browser as an anonymous user.  You should be able to view it.
  - Set video permissions to 'Only me'.  
    - You should be able to view it.  
    - Log in on another browser as someone who is on the collection's view lists (but not admin lists, not superuser).  You should not be able to view it.  
    - Log in as someone who is on the collection's admin lists.  You should be able to view it.
  - Set video permissions to 'Moira lists' and choose a list that is not in the collections view or admin lists.  
    - In another browser, log in as someone who is on that list, you should be able to view it.  
    - Log in as someone who is on one of the collection's admin lists, you should be able to view it.
    - Log in as someone who is on one of the collection's view lists, you should not be able to view it.
  - Switch the video permissions back to 'Same as collection'.  Verify that the collection's view permissions are being applied to the video.

- Set `ENABLE_VIDEO_PERMISSIONS=False` in your `.env` file.
  - The video-specific permissions options should not appear in the video edit form, and only collection-level permissions should be enforced.

Screenshot of edit video form (@roberthouse54 does this look ok to you):
screen shot 2017-11-07 at 10 38 56 am

### CODE REVIEW

File: static/js/components/dialogs/EditVideoFormDialog.js [hunk 1]
```diff
@@ -97,28 +182,95 @@ class EditVideoFormDialog extends React.Component {
...
           value={editVideoForm.title}
+          validationMessage={errors ? errors.title : ''}
         />
         <Textarea
           label="Description"
           id="video-description"
           onChange={this.setEditVideoDesc}
           value={editVideoForm.description}
         />
+        { SETTINGS.FEATURES.ENABLE_VIDEO_PERMISSIONS &&
```
Comment (gsidebo) [inline static/js/components/dialogs/EditVideoFormDialog.js]: This is a lot of code to put in a block like this. Can you move it to a helper method (`renderPermissionOptions` et al) and call that instead if the feature is true?

File: static/js/components/dialogs/EditVideoFormDialog.js [hunk 2]
```diff
@@ -47,48 +57,123 @@ class EditVideoFormDialog extends React.Component {
...
     const { dispatch } = this.props;
-    dispatch(uiActions.initEditVideoForm({
+
+    let viewChoice = video.is_private
+      ? PERM_CHOICE_NONE
+      : video.is_public
+        ? PERM_CHOICE_PUBLIC
+        : video.view_lists.length > 0
+          ? PERM_CHOICE_LISTS
+          : PERM_CHOICE_COLLECTION;
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/components/dialogs/EditVideoFormDialog.js]: Sure

File: static/js/components/dialogs/EditVideoFormDialog.js [hunk 3]
```diff
@@ -47,48 +57,123 @@ class EditVideoFormDialog extends React.Component {
...
+
+  onClose = ()=> {
+    const { hideDialog } = this.props;
+    hideDialog();
+  };
+
+  handleError = (error: Error) => {
+    const { dispatch, videoUi: { editVideoForm } } = this.props;
+    dispatch(
+      validateForm({
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/components/dialogs/EditVideoFormDialog.js]: OK

File: static/js/components/dialogs/EditVideoFormDialog.js (hunk 1 above)
Comment (mbertrand) [reply static/js/components/dialogs/EditVideoFormDialog.js]: OK

File: static/js/reducers/videoUi.js [hunk 4]
```diff
@@ -23,12 +32,41 @@ export const INITIAL_UPLOAD_SUBTITLE_FORM_STATE = {
...
+  editVideoForm: INITIAL_EDIT_VIDEO_FORM_STATE,
   videoSubtitleForm: INITIAL_UPLOAD_SUBTITLE_FORM_STATE,
-  corner: Object.keys(CANVASES)[0]
+  corner: Object.keys(CANVASES)[0],
 };
 
+
+const updateVideoForm = (state: VideoUiState, key: string, newValue: ?string) => ({
+  ...state,
+  ['editVideoForm']: {
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/reducers/videoUi.js]: Doesn't seem necessary, I'll remove them

File: static/js/reducers/videoUi.js [hunk 5]
```diff
@@ -23,12 +32,41 @@ export const INITIAL_UPLOAD_SUBTITLE_FORM_STATE = {
...
+        ...state.editVideoForm,
+        ...action.payload
+      }
+    };
+  case SET_EDIT_VIDEO_TITLE:
+    return {...state, editVideoForm: { ...state.editVideoForm, title: action.payload }};
+  case SET_EDIT_VIDEO_DESC:
+    return {...state, editVideoForm: { ...state.editVideoForm, description: action.payload }};
+  case SET_PERM_OVERRIDE_CHOICE:
+    return {...state, editVideoForm: { ...state.editVideoForm, overrideChoice: action.payload }};
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/reducers/videoUi.js]: No, updated to use it.

File: static/scss/generalStyles.scss [hunk 6]
```diff
@@ -148,3 +148,62 @@ a.button-link {
...
+.wideLabel {
+  label {
+    width: 100% !important;
+  }
+}
+
+.collectionPerms {
+  margin-left: 50px;
+  font-size: 14px;
+}
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/scss/generalStyles.scss]: OK

File: ui/permissions.py [hunk 7]
```diff
@@ -33,7 +34,32 @@ def is_staff_or_superuser(user):
...
+    if settings.ENABLE_VIDEO_PERMISSIONS:
+        if obj.is_public or request.user.is_superuser or request.user == obj.collection.owner:
+            return True
+        if obj.is_private:
+            return has_admin_permission(obj.collection, request)
+        if request.method in SAFE_METHODS:
+            view_list = list(obj.view_lists.values_list('name', flat=True))
+            if view_list:
+                return has_common_lists(request.user,
+                                        view_list + list(obj.collection.admin_lists.values_list('name', flat=True)))
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply ui/permissions.py]: Sure

**Comment by gsidebo** [comment]
I believe you can just use the pytest-django-provided 'settings' fixture
here instead of using '@override_settings':
https://pytest-django.readthedocs.io/en/latest/helpers.html#settings

**Comment by mbertrand** [comment]
great, thanks

File: ui/permissions_test.py [hunk 8]
```diff
@@ -533,6 +533,115 @@ def test_video_admin_permission_matching_lists_post(mock_moira_client,
     assert video_permission.has_object_permission(request_data.request, request_data.view, video) is True
 
 
+@pytest.mark.parametrize("enabled", [True, False])
```
Comment (gsidebo) [inline ui/permissions_test.py]: Can you globally change `enabled` to something like `video_permission_setting`? Take this line for example: `assert video_permission.has_object_permission(request_data_anon.request, request_data_anon.view, video) is enabled`. The words `is enabled` are misleading, I think. We're checking to see if two boolean values match. `assert video_permission.has_object_permission() is video_permission_setting` is a bit clearer about that than `assert video_permission.has_object_permission() is enabled`

File: ui/permissions_test.py [hunk 9]
```diff
@@ -533,6 +533,115 @@ def test_video_admin_permission_matching_lists_post(mock_moira_client,
...
+def test_override_video_private_collection_view_lists(enabled,
+                                                      settings,
+                                                      mock_moira_client,
+                                                      request_data,
+                                                      video_permission,
+                                                      video,
+                                                      moira_list):
+    """
+    A private video should not be viewable by users in the collection view lists if video permissions enabled
+    :return:
```
Comment (gsidebo) [inline ui/permissions_test.py]: Can be removed

File: static/js/components/dialogs/EditVideoFormDialog_test.js [hunk 10]
```diff
@@ -61,51 +78,121 @@ describe('EditVideoFormDialog', () => {
...
-    store.dispatch(setEditVideoDesc(newValues.description));
-    await listenForActions([
-      actions.videos.patch.requestType
-    ], () => {
-      // Calling onAccept directly b/c click doesn't work in JS tests due to MDC
-      // $FlowFixMe: Flow... come on. 'wrapper' cannot be undefined at this point.
-      wrapper.find('EditVideoFormDialog').find('Dialog').prop('onAccept')();
+    }]
+  ]) {
+    it(`updates the video when form is submitted and video permissions ${expect(videoPerms)} be enabled`, async () => {
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/components/dialogs/EditVideoFormDialog_test.js]: OK

File: static/js/components/material/Radio.js [hunk 11]
```diff
@@ -22,12 +24,14 @@ export default class Radio extends React.Component {
...
-      children
+      children,
+      className,
+      disabled
     } = this.props;
 
     let htmlId = `${radioGroupName}-${id}`;
 
-    return <div className="mdc-form-field" key={htmlId}>
+    return <div className={`mdc-form-field ${className ? className : ''}`} key={htmlId}>
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/components/material/Radio.js]: OK

File: ui/permissions_test.py (hunk 8 above)
Comment (mbertrand) [reply ui/permissions_test.py]: ok

File: ui/permissions_test.py (hunk 9 above)
Comment (mbertrand) [reply ui/permissions_test.py]: ok

File: static/js/components/material/Textfield.js [hunk 12]
```diff
@@ -10,7 +10,7 @@ export default class Textfield extends React.Component {
   };
 
   render() {
-    const { label, id, validationMessage, ...otherProps } = this.props;
+    const { label, id, validationMessage,  ...otherProps } = this.props;
```
Comment (gsidebo) [inline static/js/components/material/Textfield.js]: Extra space here

**Pull request merged**